src/Build/Graph/ProjectGraph.cs

```diff
@@ -317,10 +322,71 @@ public ProjectGraph(IEnumerable<ProjectGraphEntryPoint> entryPoints, ProjectColl
+            _allParsedProjects.TryAdd(configurationMetadata, graphNode);
             return graphNode;
         }
 
+        /// <summary>
+        /// Load a graph with root node at entryProjectFile
+        /// Maintain a queue of projects to be processed and evaluate projects in parallel
+        /// </summary>
+        private void LoadGraph(ConfigurationMetadata entryProject, ProjectCollection projectCollection, PropertyDictionary<ProjectPropertyInstance> globalProperties, ConcurrentDictionary<ConfigurationMetadata, Task> tasksInProgress)
+        {
+            var waitHandle = new AutoResetEvent(false);
+            var projectsToEvalute = new ConcurrentQueue<ConfigurationMetadata>();
+            projectsToEvalute.Enqueue(entryProject);
+            while (projectsToEvalute.Count != 0 || tasksInProgress.Any())
+            {
+                ConfigurationMetadata projectToEvaluate;
+                if (projectsToEvalute.Count != 0)
+                {
+                    projectToEvaluate = projectsToEvalute.Dequeue();
+                    var task = new Task(() =>
+                    {
+                        ProjectGraphNode parsedProject = CreateNewNode(projectToEvaluate, projectCollection);
+                        IEnumerable<ProjectItemInstance> projectReferenceItems = parsedProject.ProjectInstance.GetItems(ProjectReferenceItemName);
+                        foreach (var projectReferenceToParse in projectReferenceItems)
+                        {
+                            if (!string.IsNullOrEmpty(projectReferenceToParse.GetMetadataValue(ToolsVersionMetadataName)))
+                            {
+                                throw new InvalidOperationException(string.Format(
+                                    CultureInfo.InvariantCulture,
+                                    ResourceUtilities.GetResourceString(
+                                        "ProjectGraphDoesNotSupportProjectReferenceWithToolset"),
+                                    projectReferenceToParse.EvaluatedInclude,
+                                    parsedProject.ProjectInstance.FullPath));
+                            }
+
+                            string projectReferenceFullPath = projectReferenceToParse.GetMetadataValue(FullPathMetadataName);
+                            PropertyDictionary<ProjectPropertyInstance> projectReferenceGlobalProperties = GetProjectReferenceGlobalProperties(projectReferenceToParse, globalProperties);
+                            var projectReferenceConfigurationMetadata = new ConfigurationMetadata(projectReferenceFullPath, projectReferenceGlobalProperties);
+                            if (!_allParsedProjects.TryGetValue(projectReferenceConfigurationMetadata, out ProjectGraphNode parsedProjectReference))
+                            {
+                                if (!tasksInProgress.TryGetValue(projectReferenceConfigurationMetadata, out Task taskInProgress))
```

Comment:
I don't think the values of `tasksInProgress` are ever used. Maybe this should be a set instead of a dictionary.